arcgis-ios-sdk-samples/Layers/Style WMS layers/StyleWebMapServiceLayerViewController.swift

```diff
@@ -0,0 +1,107 @@
+    /// The map displayed in the map view.
+    let map: AGSMap
+    let layer: AGSWMSLayer
+    var styles = [String]()
+    
+    /// The map view managed by the view controller.
+    @IBOutlet weak var mapView: AGSMapView!
+    @IBOutlet weak var segmentedControl: UISegmentedControl!
+    
+    required init?(coder aDecoder: NSCoder) {
+        // Create the map.
+        map = AGSMap()
+        
+        // Create the WMS layer and add it to the map.
+        layer = AGSWMSLayer(
+            url: URL(string: "http://geoint.lmic.state.mn.us/cgi-bin/wms?VERSION=1.3.0&SERVICE=WMS&REQUEST=GetCapabilities")!,
+            layerNames: ["fsa2017"]
+        )
+        map.operationalLayers.add(layer)
+        
+        super.init(coder: aDecoder)
+        
+        // Load the WMS layer.
+        layer.load { [weak self] (error) in
+            if let error = error {
+                self?.layerDidFailToLoad(with: error)
+            } else {
+                self?.layerDidLoad()
+            }
+        }
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        (self.navigationItem.rightBarButtonItem as! SourceCodeBarButtonItem).filenames = ["StyleWebMapServiceLayerViewController"]
+        
+        // Assign the map to the map view.
+        mapView.map = map
+        
+        // The segmented control needs to be enabled here if the layer is
+        // already loaded.
+        enableSegmentedControlIfPosasible()
+    }
+    
+    /// Called in response to the layer loading successfully.
+    func layerDidLoad() {
+        guard let sublayer = layer.sublayers.firstObject as? AGSWMSSublayer else {
+            return
+        }
+        styles = sublayer.sublayerInfo.styles
+        // The segmented control needs to be enabled here if the view is already
+        // loaded.
+        enableSegmentedControlIfPosasible()
+    }
+    
+    /// Called in response to the layer failing to load. Presents an alert
+    /// announcing the failure.
+    ///
+    /// - Parameter error: The error that caused loading to fail.
+    func layerDidFailToLoad(with error: Error) {
+        let alertController = UIAlertController(title: nil, message: "Failed to load WMS layer", preferredStyle: .alert)
+        let okayAction = UIAlertAction(title: "OK", style: .default)
+        alertController.addAction(okayAction)
+        alertController.preferredAction = okayAction
+        present(alertController, animated: true)
+    }
+    
+    /// Enables the segmented control if the view has loaded and there is more
+    /// than one style.
+    func enableSegmentedControlIfPosasible() {
+        guard isViewLoaded, styles.count > 1 else { return }
+        segmentedControl.isEnabled = true
+    }
+    
+    @IBAction func changeStyle(_ sender: UISegmentedControl) {
+        guard let sublayer = layer.sublayers.firstObject as? AGSWMSSublayer else {
+            preconditionFailure()
```

Comment:
is this better than an `assert`?

Reply:
Actually, I would prefer a graceful failure like @nixta suggests. So that the app published to the store won't bomb out.

Reply:
@esreli Good question. The advantage of `precondition()`/`preconditionFailure()` is twofold:
1. It prevents forward progress in both debug (non-optimized) and release (optimized) builds.
2. It returns `Never`. If this were an assert, a return statement would also be needed.

`preconditionFailure()` is a good choice here because we already know that the layer exists. If the layer didn't exist, then there wouldn't be any styles and the segmented control wouldn't be enabled. The fact that the handler fired is proof that the the sublayer does exist. If the control somehow became enabled without getting any styles, we have bigger problems and should bail immediately.

I could have force unwrapped `firstObject`, but that leaves ambiguity about whether the `nil` case has been considered. Using the `guard`/`preconditionFailure()` combo makes it explicit that I have considered the `nil` case and essentially deemed it impossible.

I also could have subscripted into `sublayers`, but that really isn't any different than `firstObject!`.

If you'd like, I could add a message to make it even more explicit why we can't move forward:
`preconditionFailure("Styles control enabled without a sublayer")`

Reply:
For an app that is published to the iTunes App store, I feel strongly about removing a forced crash for a potentially missing web asset layer. 

In the event we have bigger problems, I would encourage it be handled gracefully instead of bailing. I would rather the app's component fail than the entire app fail.

Reply:
> If the control somehow became enabled without getting any styles, we have bigger problems and should bail immediately.

Then I would ask that we either gracefully log something from the `guard` and not kill the app if we feel that this is application logic error, or that we `if let` our way to a `sublayer` variable if we feel that it'll never rightly happen.

And should we not rather/also be guarding that the following `selectedSegmentIndex` is within range for the `styles` array?

More fundamentally, I question the use of a `guard` statement whose content kills the app. `guard` is meant to help defend against issues, not trigger them (I'd consider a hard kill an issue, whether it's us or the array access pulling the trigger). By all means use a guard statement to ensure that the next bits of code will work, but surely you don't want to make the result of them not working terminal for the whole app?